src/libcore/ptr.rs

```diff
@@ -946,6 +946,58 @@ pub unsafe fn write_volatile<T>(dst: *mut T, src: T) {
     intrinsics::volatile_store(dst, src);
 }
 
+/// Freezes `count * size_of::<T>()` bytes of memory, converting undefined data into
+/// arbitrary but fixed data.
+///
+/// Uninitialized memory has undefined contents, and interation with that data
+/// can easily cause undefined behavior. This function "freezes" memory
+/// contents, converting uninitialized memory to initialized memory with
+/// arbitrary contents so that use of it is well defined.
+///
+/// This function has no runtime effect; it is purely an instruction to the
+/// compiler. In particular, it does not actually write anything to the memory.
+///
+/// # Safety
+///
+/// Behavior is undefined if any of the following conditions are violated:
+///
+/// * `dst` must be [valid] for writes.
+///
+/// * Every bit representation of `T` must be a valid value.
+///
+/// Note that even if `T` has size `0`, the pointer must be non-NULL and properly aligned.
+///
+/// [valid]: ../ptr/index.html#safety
+///
+/// # Examples
+///
+/// ```ignore (io-is-in-std)
+/// use std::io::{self, Read};
+/// use std::mem;
+/// use std::ptr;
+///
+/// pub fn read_le_u32<R>(reader: &mut R) -> io::Result<u32>
+/// where
+///     R: Read,
+/// {
+///     unsafe {
+///         // We're passing this buffer to an arbitrary reader and aren't
+///         // guaranteed they won't read from it, so freeze to avoid UB.
+///         let mut buf: [u8; 4] = mem::uninitialized();
+///         ptr::freeze(&mut buf, 1);
+///         reader.read_exact(&mut buf)?;
+///
+///         Ok(u32::from_le_bytes(buf))
+///     }
+/// }
+/// ```
+#[inline]
+#[unstable(feature = "ptr_freeze", issue = "0")]
+pub unsafe fn freeze<T>(dst: *mut T, count: usize) {
+    let _ = count;
+    asm!("" : "=*m"(dst) : );
```

Comment:
The constraints here are incorrect since they only tell LLVM that the inline asm is writing to **one** element of `dst`.

What you want is something like this: `asm!("" :: "r" (dst) : "memory" : "volatile");`

Reply:
Also I just noticed that the current constraints indicate that the inline asm is overwriting the value in `dst` without reading it, which means that LLVM is free to discard any previous writes to `dst` which might have initialized it to a particular value.

Reply:
Why would someone be using this function if they've already initialized the memory?

Reply:
> What you want is something like this: `asm!("" :: "r" (dst) : "memory" : "volatile");`

The `memory` clobber handles all memory, right? Why do we need to pass in `dst` in that case?

Reply:
> Why would someone be using this function if they've already initialized the memory?

The docs for this function say that this function has no runtime effect. I would assume that this means that already-initialized data passed to this function would remain untouched.

> The `memory` clobber handles all memory, right? Why do we need to pass in `dst` in that case?

The `memory` clobber tells LLVM to assume that any memory *whose address might be visible to other threads* has been modified. We pass the address to the inline asm so that LLVM must assume the address is globally visible.

Reply:
"Clobbers all memory" does not mean literally all memory in naive source language semantics, e.g. certainly in `let x = 1; asm!(... : "memory" : "volatile"); print(x);` the variable `x` will be optimized out. It's nebulous (to me at least) what the rules are but it's definitely safer to make the pointer actually available to the asm block.

Reply:
> Why would someone be using this function if they've already initialized the memory?

First of all, you called it "freeze", and that is what LLVM's proposed "freeze" does: pick non-deterministic values for uninitialized data but keep initialized data intact.

Second, we have a use-case for that in https://github.com/crossbeam-rs/crossbeam/issues/315.